ansible_base/rbac/api/serializers.py

```diff
@@ -0,0 +1,198 @@
+            for cls in permission_registry.all_registered_models
+        ]
+
+    def get_dynamic_object(self, data):
+        model = data.rsplit('.')[-1]
+        return permission_registry.content_type_model.objects.get(model=model)
+
+    def to_representation(self, value):
+        return f'{settings.ANSIBLE_BASE_SERVICE_PREFIX}.{value.model}'
+
+
+class PermissionField(ChoiceLikeMixin):
+    psuedo_field = 'codename'
+
+    def get_model_for_init(self):
+        return permission_registry.permission_model
+
+    def get_dynamic_choices(self):
+        perms = []
+        for cls in permission_registry.all_registered_models:
+            cls_name = cls._meta.model_name
+            for action in cls._meta.default_permissions:
+                perms.append(f'{settings.ANSIBLE_BASE_SERVICE_PREFIX}.{action}_{cls_name}')
+            for perm_name, description in cls._meta.permissions:
+                perms.append(f'{settings.ANSIBLE_BASE_SERVICE_PREFIX}.{perm_name}')
+        return perms
+
+    def get_dynamic_object(self, data):
+        codename = data.rsplit('.')[-1]
+        return permission_registry.permission_model.objects.get(codename=codename)
+
+    def to_representation(self, value):
+        return f'{settings.ANSIBLE_BASE_SERVICE_PREFIX}.{value.codename}'
+
+
+class ManyRelatedListField(serializers.ListField):
+    def to_representation(self, data):
+        "Adds the .all() to treat the value as a queryset"
+        return [self.child.to_representation(item) if item is not None else None for item in data.all()]
+
+
+class RoleDefinitionSerializer(CommonModelSerializer):
+    # Relational versions - we may switch to these if custom permission and type models are exposed but out of scope here
+    # permissions = serializers.SlugRelatedField(many=True, slug_field='codename', queryset=permission_registry.permission_model.objects.all())
+    # content_type = ContentTypeField(slug_field='model', queryset=permission_registry.content_type_model.objects.all(), allow_null=True, default=None)
+    permissions = ManyRelatedListField(child=PermissionField())
+    content_type = ContentTypeField(allow_null=True, default=None)
+
+    class Meta:
+        model = RoleDefinition
+        fields = '__all__'
+
+    def validate(self, validated_data):
+        validate_permissions_for_model(validated_data.get('permissions', []), validated_data.get('content_type'))
+        return super().validate(validated_data)
+
+
+class RoleDefinitionDetailSeraizler(RoleDefinitionSerializer):
+    content_type = ContentTypeField(read_only=True)
+
+
+class BaseAssignmentSerializer(CommonModelSerializer):
+    content_type = ContentTypeField(read_only=True)
+
+    def create(self, validated_data):
+        rd = validated_data['role_definition']
+        model = rd.content_type.model_class()
+        obj = model.objects.get(id=validated_data['object_id'])
```

Comment:
This breaks when the PK field name is not `id`. Pulp uses `pulp_id` for example.

```
pulp-1          |   File "/src/django-ansible-base/ansible_base/rbac/api/serializers.py", line 154, in create
pulp-1          |     obj = model.objects.get(id=validated_data['object_id'])
pulp-1          |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/manager.py", line 87, in manager_method
pulp-1          |     return getattr(self.get_queryset(), name)(*args, **kwargs)
pulp-1          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 623, in get
pulp-1          |     clone = self._chain() if self.query.combinator else self.filter(*args, **kwargs)
pulp-1          |                                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 1436, in filter
pulp-1          |     return self._filter_or_exclude(False, args, kwargs)
pulp-1          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 1454, in _filter_or_exclude
pulp-1          |     clone._filter_or_exclude_inplace(negate, args, kwargs)
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 1461, in _filter_or_exclude_inplace
pulp-1          |     self._query.add_q(Q(*args, **kwargs))
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1546, in add_q
pulp-1          |     clause, _ = self._add_q(q_object, self.used_aliases)
pulp-1          |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1577, in _add_q
pulp-1          |     child_clause, needed_inner = self.build_filter(
pulp-1          |                                  ^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1427, in build_filter
pulp-1          |     lookups, parts, reffed_expression = self.solve_lookup_type(arg, summarize)
pulp-1          |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1237, in solve_lookup_type
pulp-1          |     _, field, _, lookup_parts = self.names_to_path(lookup_splitted, self.get_meta())
pulp-1          |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1725, in names_to_path
pulp-1          |     raise FieldError(
pulp-1          | django.core.exceptions.FieldError: Cannot resolve keyword 'id' into field. Choices are: alternatecontentsourcepath, ansible_crossrepositorycollectionversionindex, content, core_pulp_exporter, description, distributions, download_logs, gpgkey, group_roles, last_synced_metadata_time, name, next_version, private, pulp_created, pulp_domain, pulp_domain_id, pulp_id, pulp_labels, pulp_last_updated, pulp_type, pulpimporterrepository, remote, remote_id, repositories, repository_ptr, repository_ptr_id, repositorycontent, retain_repo_versions, uploads, upstream_repositories, user_hidden, user_roles, versions
```

Reply:
there are a few other places where `content_object.id` is used as well.